http-mclient/src/Network/HTTP/MClient.hs

```diff
@@ -1,156 +1,131 @@
-  deriving (Conversion Text) via Text
-  deriving stock (Eq, Show)
-
-data HttpError =
-    ConnectionError HTTP.HttpException
-  | DecodeFailure DecodeError Response
-  | InvalidStatusCode StatusCode Response
-  | InvalidContentType Response
-  | UnsupportedContentType MediaType Response
-  deriving stock (Show)
-
-instance Eq HttpError where
-  DecodeFailure t r          == DecodeFailure t' r'          = t == t' && r == r'
-  InvalidStatusCode code r   == InvalidStatusCode code' r'   = r == r' && code == code'
-  InvalidContentType r       == InvalidContentType r'        = r == r'
-  UnsupportedContentType m r == UnsupportedContentType m' r' = m == m' && r == r'
-  _                          == _                            = False
-
-instance Exception HttpError
-
-data ContentType
-  = FormUrlEncoded
-  | Json
-  | PlainText
-  | XML
-
-class HasMediaType (ctyp :: ContentType) where
-  mediaType :: MediaType
-  mediaType = NE.head (mediaTypes @ctyp)
-
-  mediaTypes :: NE.NonEmpty MediaType
-  mediaTypes = mediaType @ctyp NE.:| []
-
-instance HasMediaType 'FormUrlEncoded where
-  mediaType = "application" // "x-www-form-urlencoded"
-
-instance HasMediaType 'Json where
-  mediaType = "application" // "json" /: ("charset", "utf-8")
-
-instance HasMediaType 'PlainText where
-  mediaTypes
-    = "text" // "plain" /: ("charset", "utf-8")
-    NE.:|
-    [ "text" // "html" /: ("charset", "utf-8")]
-
-mkRequest
-  :: forall contentType acceptsType e a
-  . ( ToText e
-    , HasMediaType contentType
-    , HasMediaType acceptsType
-    )
-  => SendRequest
-  -> (LBS.ByteString -> Either e a)
-  -> HTTP.Request
-  -> IO (Either HttpError a)
-mkRequest sendRequest decodeBody =
-  runExceptT . mkRequest' @contentType @acceptsType (fromType @200) sendRequest decodeBody
-
-mkRequest'
-  :: forall contentType acceptsType a e
-  . ( ToText e
-    , HasMediaType contentType
-    , HasMediaType acceptsType
-    )
-  => StatusCode
-  -> SendRequest
-  -> (LBS.ByteString -> Either e a)
+data ResponseError
+  = BodyDecodeFailure Text
+  | HTTPError HTTP.HttpExceptionContent
+  | UnexpectedContentType BS.ByteString
+  | MissingContentType
+  | UnexpectedStatusCode HTTP.Status
+  deriving stock Show
+
+instance Exception.Exception ResponseError
+
+type Result a = Either ResponseError a
+
+type ResponseDecoder a = HTTP.Response LBS.ByteString -> Result a
+
+decodeStatus
+  :: HTTP.Status
+  -> ResponseDecoder a
+  -> ResponseDecoder a
+decodeStatus expectedStatus decoder = decodeStatuses [(expectedStatus, decoder)]
+
+decodeStatuses
+  :: [(HTTP.Status, ResponseDecoder a)]
+  -> ResponseDecoder a
+decodeStatuses decoders response
+  = maybe (Left $ UnexpectedStatusCode statusCode) ($ response)
+  $ List.lookup statusCode decoders
+  where
+    statusCode = HTTP.responseStatus response
+
+decodeContentType :: BS.ByteString -> ResponseDecoder a -> ResponseDecoder a
+decodeContentType accepted decoder = decodeContentTypes [(accepted, decoder)]
```

Comment:
Is the choice of `BS.ByteString` to prevent unnecessary type conversions from Text. I am assuming Text would have been an ideal candidate for the `content-type` header.